### PR TITLE
Messages referenced

### LINKED ISSUE #174: Client - MQTT messages/referenced subscription
## Description

Implements the `messages/referenced` event API.

### PR DESCRIPTION
**Pull request by rebinmannady93**
# Description of change

fixes #174 

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

test_messages_referenced_event.c

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests using CTest that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: examples/get_event_messages.c [hunk 1]
```diff
@@ -46,12 +48,31 @@ void callback(event_client_event_t *event) {
...
+          printf("Parent Id %zu : %s\n", i + 1, res_msg_metadata_parent_get(res, i));
+        }
+        printf("Inclusion State : %s\n", res->inclusion_state);
+        printf("Is Solid : %s\n", res->is_solid ? "true" : "false");
+        printf("Should Promote : %s\n", res->should_promote ? "true" : "false");
+        printf("Should Reattach : %s\n", res->should_reattach ? "true" : "false");
+        printf("Referenced Milestone : %ld\n", res->referenced_milestone);
+      }
+      res_msg_metadata_free(res);
+    }
```
Comment (oopsmonk) [inline examples/get_event_messages.c:80]: is there a proper way to handle errors?
we can simply clean up and terminate the program when errors occur.

File: src/client/api/events/sub_messages_metadata.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+/**
+ * @brief Frees a message metadata reponse object
+ * @param[in] res A response object
+ */
+void res_msg_metadata_free(msg_metadata_t *res);
+
+/**
+ * @brief Gets the number of parent messages
+ *
+ * @param res The message metadata response
```
Comment (oopsmonk) [inline src/client/api/events/sub_messages_metadata.h]:
```suggestion
 * @param[in] res The message metadata response
```

File: src/client/api/events/sub_messages_metadata.h [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+ * @return size_t The number of parent message ids
+ */
+size_t res_msg_metadata_parents_len(msg_metadata_t *res);
+
+/**
+ * @brief Gets the parent message ID by a given index
+ *
+ * @param res The message metadata respose
+ * @param index A index
```
Comment (oopsmonk) [inline src/client/api/events/sub_messages_metadata.h]:
```suggestion
 * @param[in] res The message metadata response
 * @param[in] index An index of parent ID
```

File: src/client/api/events/sub_messages_metadata.h [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+ * @param[out] res Parsed response object
+ * @return 0 if success
+ */
+int parse_messages_metadata(char *data, msg_metadata_t *res);
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif
```
Comment (oopsmonk) [inline src/client/api/events/sub_messages_metadata.h]: fix git diff: `no newline at end of file`.
FYI. https://thoughtbot.com/blog/no-newline-at-end-of-file
```suggestion
#endif

```

File: tests/client/api_events/test_messages_events.c [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+int main() {
+  UNITY_BEGIN();
+
+  RUN_TEST(test_messages_metadata_parser);
+  RUN_TEST(test_messages_referenced_events);
+
+  return UNITY_END();
+}
```
Comment (oopsmonk) [inline tests/client/api_events/test_messages_events.c]: as mentioned previously, newline

File: tests/client/api_events/test_milestones_confirmed_event.c [hunk 6]
```diff
@@ -72,6 +75,9 @@ void test_milestones_confirmed_events(void) {
   TEST_ASSERT_NOT_NULL(client);
   TEST_ASSERT_EQUAL_INT(0, event_register_cb(client, &callback));
   TEST_ASSERT_EQUAL_INT(0, event_start(client));
+  // Wait until test is completed
+  while (!test_completed)
```
Comment (oopsmonk) [inline tests/client/api_events/test_milestones_confirmed_event.c]: if no event was received by the MQTT client, this test case will be an infinite loop.
maybe add a timeout?

**Review by oopsmonk — CHANGES_REQUESTED** [review]

File: src/client/api/events/sub_messages_metadata.h (hunk 2 above)
Comment (rebinmannady93) [reply src/client/api/events/sub_messages_metadata.h]: Done.

File: src/client/api/events/sub_messages_metadata.h (hunk 3 above)
Comment (rebinmannady93) [reply src/client/api/events/sub_messages_metadata.h]: Done.

File: src/client/api/events/sub_messages_metadata.h (hunk 4 above)
Comment (rebinmannady93) [reply src/client/api/events/sub_messages_metadata.h]: Done.

File: tests/client/api_events/test_messages_events.c (hunk 5 above)
Comment (rebinmannady93) [reply tests/client/api_events/test_messages_events.c]: Done.

File: examples/get_event_messages.c (hunk 1 above)
Comment (rebinmannady93) [reply examples/get_event_messages.c:80]: Done.

File: tests/client/api_events/test_milestones_confirmed_event.c (hunk 6 above)
Comment (rebinmannady93) [reply tests/client/api_events/test_milestones_confirmed_event.c]: Done.

**Review by oopsmonk — APPROVED** [review]

**Pull request merged**